kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/codgen/GeneratedAdaptersTest.kt

```diff
@@ -1155,6 +1156,40 @@ class GeneratedAdaptersTest {
     assertThat(decoded).isEqualTo(HasCollectionOfPrimitives(listOf(4, -5, 6)))
   }
 
+  @JsonClass(generateAdapter = true, generator = "custom")
+  data class CustomGeneratedClass(val foo: String)
+
+  @Test fun customGenerator_withClassPresent() {
+    val moshi = Moshi.Builder().build()
+    val adapter = moshi.adapter(CustomGeneratedClass::class.java)
+    val unwrapped = (adapter as NullSafeJsonAdapter<CustomGeneratedClass>).delegate()
+    assertThat(unwrapped).isInstanceOf(GeneratedAdaptersTest_CustomGeneratedClassJsonAdapter::class.java)
+  }
+
+  @JsonClass(generateAdapter = true, generator = "custom")
+  data class CustomGeneratedClassMissing(val foo: String)
+
+  @Test fun customGenerator_withClassMissing() {
+    val moshi = Moshi.Builder().build()
+    try {
+      moshi.adapter(CustomGeneratedClassMissing::class.java)
+      fail()
+    } catch (e: RuntimeException) {
+      assertThat(e).hasMessageContaining("Failed to find the generated JsonAdapter class")
+    }
+  }
+}
+
+// Has to be outside to avoid Types seeing an owning class
+@JsonClass(generateAdapter = true)
```

Comment:
I don't know why this shows up as changed, I didn't change it in the branch but it shows up here after rebasing in the github GUI since I didn't have any conflicts locally